src/mlpack/methods/ann/layer/batch_norm_impl.hpp

```diff
@@ -0,0 +1,373 @@
+    throw std::invalid_argument("BatchNormType::CustomInitialize(): wrong "
+        "rows size!"); 
+  }
+  MatType gammaTemp;
+  MatType betaTemp;
+  // Gamma acts as the scaling parameters for the normalized output.
+  MakeAlias(gammaTemp, W.memptr(), size, 1);
+  // Beta acts as the shifting parameters for the normalized output.
+  MakeAlias(betaTemp, W.memptr() + gammaTemp.n_elem, size, 1);
+
+  gammaTemp.fill(1.0);
+  betaTemp.fill(0.0);
+
+  runningMean.zeros(size, 1);
+  runningVariance.ones(size, 1);
+}
+
+template<typename MatType>
+void BatchNormType<MatType>::Forward(
+    const MatType& input,
+    MatType& output)
+{
+  const size_t batchSize = input.n_cols;
+  const size_t inputSize = inputDimension;
+
+  // Set size of output equal to the size of input.
+  // output.set_size(arma::size(input));
+
+  // We will calculate minibatch norm on each channel / feature map.
+  if (this->training)
+  {
+    // Check only during training, batch-size can be one during inference.
+    if ((batchSize * higherDimension) == 1 && inputSize == 1)
+    {
+      Log::Warn << "Variance for single element isn't defined and" <<
+          " will be set to  0.0 for training. Use a batch-size" <<
+          " greater than 1 to fix the warning." << std::endl;
+    }
+
+    // Input corresponds to output from convolution layer.
+    // Use a cube for simplicity.
+    arma::Cube<typename MatType::elem_type> inputTemp(
+        const_cast<MatType&>(input).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+
+    // Initialize output to same size and values for convenience.
+    arma::Cube<typename MatType::elem_type> outputTemp(
+        const_cast<MatType&>(output).memptr(), inputSize, size,
+        batchSize * higherDimension, false, false);
+    outputTemp = inputTemp;
+
+    // Calculate mean and variance over all channels.
+    mean = arma::mean(arma::mean(inputTemp, 2), 0);
+    variance = arma::mean(arma::mean(arma::pow(
+        inputTemp.each_slice() - arma::repmat(mean,
+        inputSize, 1), 2), 2), 0);
+
+    outputTemp.each_slice() -= arma::repmat(mean, inputSize, 1);
+
+    // Used in backward propagation.
+    inputMean.set_size(arma::size(inputTemp));
+    inputMean = outputTemp;
+
+    // Normalize output.
+    outputTemp.each_slice() /= arma::sqrt(arma::repmat(variance,
+        inputSize, 1) + eps);
+
+    // Re-used in backward propagation.
+    normalized.set_size(arma::size(inputTemp));
+    normalized = outputTemp;
+
+    outputTemp.each_slice() %= arma::repmat(gamma.t(),
+        inputSize, 1);
+    outputTemp.each_slice() += arma::repmat(beta.t(),
+        inputSize, 1);
+
+    count += 1;
+    averageFactor = average ? 1.0 / count : momentum;
+
+    double nElements = 0.0;
+    if (input.n_elem - size != 0)
+      nElements = 1.0 / (input.n_elem - size + eps);
+
+    // Update running mean and running variance.
+    runningMean = (1 - averageFactor) * runningMean + averageFactor *
+        mean.t();
```

Comment:
Is the transpose necessary here?  I am not sure I follow why.

Reply:
Mean shape is (1, size). But runningMean shape is (size, 1).